src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -162,17 +162,14 @@ ExecuteResponse execute() throws IOException {
 
       try {
         Iterator<Operation> operationStream = executionBlockingStubSupplier.get().execute(request);
-
-        // We don't want to reset backoff for Execute() since if there is an error:
-        //   1. If happened before we received a first response, we want to ensure the retry counter
-        //      is increased and call Execute() again.
-        //   2. Otherwise, we will fallback to WaitExecution() loop.
-        //
-        // This also prevent us from being stuck at a infinite loop:
-        //   Execute() -> WaitExecution() -> Execute()
-        return handleOperationStream(operationStream, /* resetWaitExecutionBackoff */ false);
+        return handleOperationStream(operationStream);
       } catch (StatusRuntimeException e) {
-        if (lastOperation != null) {
+        // If lastOperation is not null, we know the execution request is accepted by the server. In
+        // this case, we will fallback to WaitExecution() loop when the stream is broken.
+        //
+        // However, we only do so if waitExecutionBackoff should retry. Also increase the retry
+        // counter at the same time (done by nextDelayMillis()).
+        if (lastOperation != null && waitExecutionBackoff.nextDelayMillis() >= 0) {
```

Comment:
Why do you increase the delay here?

Reply:
This is internal behavior of ProgressiveBackoff: once you reset it, the next call to `nextDelayMillis()` will not increase the counter. If `lastOperation != null`, we know that the `waitExecutionBackoff` is just reset, so call `nextDelayMillis()` here to ensure the expected retry times(--remote_retry + 1).

Reply:
(`nextDelayMillis()` is used by retrier to determine whether it should retry and if so, how long to wait before try again.)